### PR TITLE
Simplify render path

### PR DESCRIPTION
**Pull request by kkaefer**
We're currently have a pretty complicated way of rendering a tile, going through many layers of objects and some double dispatching along the way. This makes the code hard to follow and we need to traverse the hierarchy several times, for rendering opaque and translucent objects. We also want to add an additional step before that, giving renderable objects (=buckets) an opportunity to upload data to the GPU before we kick off rendering a frame.

### CODE REVIEW

**Comment by kkaefer** [comment]
@mapbox/gl Please review.

**Comment by ansis** [comment]
Is there a measurable performance improvement from wrapping all calls that modify gl state?
What performance impact did the switch to uploading at the beginning of a frame have?
Any plans to port this to -js?

**Comment by kkaefer** [comment]
This fixes #434
